File: codejam/cpp/2011/1A/1A_Pseudominion.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pi;
#define FOR(i,a) for (ll i = 0; i < (a); i++)
#define len(x) (ll) x.size()
const ll INF = 1LL << 62;
const ll MOD = 1000000007;
//const ll MOD = 998244353;
const double PI = 4*atan(double(1.0));
vector<vector<ll>> twodi(ll N, ll M, ll v) { vector<vector<ll>> res(N); for (auto &vv : res) vv.resize(M,v); return res; }
struct state { ll t,nc,tidx,c1idx,c2idx; };
bool operator<(const state a, const state b) { 
    return a.t < b.t || a.t == b.t && (
           a.nc < b.nc || a.nc == b.nc && ( 
           a.tidx < b.tidx || a.tidx == b.tidx && (
           a.c1idx < b.c1idx || a.c1idx == b.c1idx && a.c2idx < b.c2idx)));
}
struct card { ll idx,c,s,t; };
bool operator<(const card &a, const card &b) { return a.s < b.s; } 
ll c0sum[81][81];

int main (int argc, char **argv) {
    if (argc > 1) { freopen(argv[1],"r",stdin); }
    ios_base::sync_with_stdio(false);cin.tie(0);
    // PROGRAM STARTS HERE
    ll T; cin >> T;
    for (ll tt=1;tt<=T;tt++) {
        vector<card> cards;
        ll N; cin >> N;
        FOR(i,N) { ll c,s,t; cin >> c >> s >> t; cards.push_back(card{i,c,s,t}); }
        ll M; cin >> M;
        FOR(i,M) { ll c,s,t; cin >> c >> s >> t; cards.push_back(card{i+N,c,s,t}); }
        vector<card> T,C0,C1,C2;
        for (auto &c : cards) {
            if (c.t > 0) T.push_back(c); else if (c.c == 0) C0.push_back(c); else if (c.c == 1) C1.push_back(c); else C2.push_back(c);
        }
        FOR(j,81) {
            vector<card> ca;
            for (auto &c : C0) { if (c.idx < j) ca.push_back(c); }
            sort(ca.begin(),ca.end());
            reverse(ca.begin(),ca.end());
            ll s = 0;
            c0sum[0][j] = 0; FOR(i,ca.size()) { s += ca[i].s; c0sum[i+1][j] = s; }
            for (ll i = ca.size()+1;i<=80;i++) { c0sum[i][j] = c0sum[ca.size()][j]; }
        }
        map<state,ll> dict;
        function<void(state)> dfs;
        dfs = [&](state s) {
            if (dict.count(s) > 0) return;
            if (s.t == 0) { dict[s] = 0; return; }
            ll v = 0;
            if (s.tidx+1 < T.size() && T[s.tidx+1].idx < s.nc) {
                card cc = T[s.tidx+1];
                state ns = state{s.t-1+cc.t,min(N+M,s.nc+cc.c),s.tidx+1,s.c1idx,s.c2idx};
                dfs(ns); v = dict[ns]+cc.s;
            } else {
                v = c0sum[min(80LL,s.t)][s.nc];
                if (s.c2idx+1 < C2.size() && C2[s.c2idx+1].idx < s.nc) {
                    card cc = C2[s.c2idx+1];
                    state ns1 = state{s.t,s.nc,s.tidx,s.c1idx,s.c2idx+1};
                    state ns2 = state{s.t-1+cc.t,min(N+M,s.nc+cc.c),s.tidx,s.c1idx,s.c2idx+1};
                    dfs(ns1); v = max(v,dict[ns1]);
                    dfs(ns2); v = max(v,cc.s+dict[ns2]);
                } else if (s.c1idx+1 < C1.size() && C1[s.c1idx+1].idx < s.nc) {
                    card cc = C1[s.c1idx+1];
                    state ns1 = state{s.t,s.nc,s.tidx,s.c1idx+1,s.c2idx};
                    state ns2 = state{s.t-1+cc.t,min(N+M,s.nc+cc.c),s.tidx,s.c1idx+1,s.c2idx};
                    dfs(ns1); v = max(v,dict[ns1]);
                    dfs(ns2); v = max(v,cc.s+dict[ns2]);
                }
            }
            dict[s] = v;
        };
        state start = state{1,N,-1,-1,-1};
        dfs(start);
        ll ans = dict[start];
        printf("Case #%lld: %lld\n",tt,ans);
    }
}

File: codejam/python/2011/1A/1A_Pseudominion.py
import sys
from collections import deque, namedtuple

## Input crap
infile = sys.stdin
intokens = deque()
def getTokens(): 
    while not intokens:
        for c in infile.readline().rstrip().split() :
            intokens.append(c)    
def gs(): getTokens(); return intokens.popleft()
def gi(): return int(gs())
def gf(): return float(gs())
def gbs(): return [c for c in gs()]
def gis(n): return [gi() for i in range(n)]
def ia(m): return [0] * m
def iai(m,v): return [v] * m
def twodi(n,m,v): return [iai(m,v) for i in range(n)]
def fill2(m) : r = gis(2*m); return r[0::2],r[1::2]
def fill3(m) : r = gis(3*m); return r[0::3],r[1::3],r[2::3]
def fill4(m) : r = gis(4*m); return r[0::4],r[1::4],r[2::4],r[3::4]
MOD = 998244353
##MOD = 1000000007
Card = namedtuple("Card","idx c s t")
MyState = namedtuple("MyState","t nc tidx c1idx c2idx")
def main() :
    if len(sys.argv) > 1 : global infile; infile = open(sys.argv[1],'rt')
    ## PROGRAM STARTS HERE
    T = gi()
    for tt in range(1,T+1) :
        T,C0,C1,C2 = [],[],[],[]
        def placecard(c) :
            if c.t > 0 : T.append(c); return
            if c.c == 0 : C0.append(c); return
            if c.c == 1 : C1.append(c); return
            C2.append(c); return
        N = gi()
        for i in range(N) : c,s,t = gi(),gi(),gi(); cc = Card(i,c,s,t); placecard(cc)
        M = gi()
        for i in range(M) : c,s,t = gi(),gi(),gi(); cc = Card(N+i,c,s,t); placecard(cc)
        ## Presolve the C0 sums
        c0sum = [[0]*81 for _ in range(81)]
        for j in range(81) :
            ca = [x for x in C0 if x.idx < j]
            ca.sort(key=lambda x: x.s)
            ca.reverse()
            s = 0
            for i,c in enumerate(ca) : s += c.s; c0sum[i+1][j] = s
            for i in range(len(ca)+1,81) : c0sum[i][j] = c0sum[len(ca)][j]
        ## Now for the poor man's DFS
        cache = {}
        startst = MyState(1,N,-1,-1,-1)
        stack = [startst]
        while stack :
            st = stack.pop()
            if st.t == 0 : cache[st] = 0; continue
            if st in cache : continue
            ## Check for the states that we need presolved to resolve this state
            if st.tidx+1 < len(T) and T[st.tidx+1].idx < st.nc :
                tidx = st.tidx+1; cc = T[tidx]
                ns = MyState(st.t-1+cc.t,min(N+M,st.nc+cc.c),tidx,st.c1idx,st.c2idx)
                if ns not in cache :
                    stack.append(st)
                    stack.append(ns)
                    continue
                else :
                    cache[st] = cc.s + cache[ns]
            else :
                score = c0sum[min(80,st.t)][st.nc]
                if st.c2idx+1 < len(C2) and C2[st.c2idx+1].idx < st.nc :
                    c2idx = st.c2idx+1; cc = C2[c2idx]
                    ns1 = MyState(st.t,st.nc,st.tidx,st.c1idx,c2idx)
                    ns2 = MyState(st.t-1+cc.t,min(N+M,st.nc+cc.c),st.tidx,st.c1idx,c2idx)
                    if ns1 not in cache or ns2 not in cache :
                        stack.append(st)
                        if ns1 not in cache : stack.append(ns1)
                        if ns2 not in cache : stack.append(ns2)
                        continue
                    else :
                        score = max(score,cache[ns1],cc.s+cache[ns2])
                elif st.c1idx+1 < len(C1) and C1[st.c1idx+1].idx < st.nc :
                    c1idx = st.c1idx+1; cc = C1[c1idx]
                    ns1 = MyState(st.t,st.nc,st.tidx,c1idx,st.c2idx)
                    ns2 = MyState(st.t-1+cc.t,min(N+M,st.nc+cc.c),st.tidx,c1idx,st.c2idx)
                    if ns1 not in cache or ns2 not in cache :
                        stack.append(st)
                        if ns1 not in cache : stack.append(ns1)
                        if ns2 not in cache : stack.append(ns2)
                        continue
                    else :
                        score = max(score,cache[ns1],cc.s+cache[ns2])
                ##print(f"DBG: st:{st} score:{score}")
                cache[st] = score
        ans = cache[startst]
        print(f"Case #{tt}: {ans}")

if __name__ == "__main__" :
    main()

